aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -181,61 +181,79 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
+  //          cause confusion for AutogradOther. It's pretty straightforward to use Autograd (if available)
+  //          in this case.
+  //    (2.3) Use kernel from DispatchKey::Autograd if available
+  //    (2.4) Special logic to handle catchAll for Autograd keys
   //          For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
   //          if there's no direct registration to the backend key.
   //          Tensor factory functions used to have no registration to Autograd key but only to catchAll.
   //          In the past we directly call into backends(filled with catchAll) after BackendSelect.
   //          Now that we first call Autograd backend keys after BackendSelect, we should fill those
   //          with catchAll as well.
-  //    The implementation of (2.1) & (2.3) relies on the invariant that for a given backend,
+  //    The implementation of (2.2) & (2.4) relies on the invariant that for a given backend,
   //    `computeDispatchTableEntryWithDebug()` will be called for that backend's autograd key after the
   //    backend key. See Note [Refresh Runtime Autograd entries in dispatchTable_]
   //  (3) Use fallthrough kernel that are registered as fallback.
   //  (4) Use catchAll kernel if available
   // Alias Key Precedence:
-  //   Math > Autograd
+  //   DefaultBackend > Math > Autograd
+  // Note [DefaultBackend and Math]
+  //   When there're registrations to both DefaultBackend & Math & Autograd, from (2.2) we know DefaultBackend
+  //   and Autograd kernels will be picked up and Math is overriden.
+  //   This is fine and in practice DefaultBackend and Math shouldn't co-exist for an op.
   // TODO: Update alias key precedence after we add new alias keys AutogradDispatchCPUOrCUDA .
-  // TODO: we can remove (2.3) and (4) after TypeDefault registrations are moved from catchAll to Math
+  // TODO: we can remove (2.4) and (4) after TypeDefault registrations are moved from catchAll to Math
   //       so that Math can populate to Autograd backend keys before fallback kernels.
 
   // 1. Operator registration
   if (auto direct_registration = getKernelForDispatchKey(dispatch_key)) {
     return {*direct_registration.value(), "kernel"};
   }
 
+  // 2.1 Use DefaultBackend kernel if available.
+  if (isIncludedInAlias(dispatch_key, DispatchKey::DefaultBackend)) {
+    if (auto default_backend_registration = getKernelForDispatchKey(DispatchKey::DefaultBackend)) {
+      return {*default_backend_registration.value(), "default backend kernel"};
+    }
+  }
+
   bool is_autograd_key_with_backend_kernel =
-    hasKernelForDispatchKeySet(getBackendKeySetFromAutograd(dispatch_key));
-  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
-  //      when there's no direct registration to its corresponding backend key.
+    hasKernelForAnyDispatchKey(getBackendKeySetFromAutograd(dispatch_key).add(DispatchKey::DefaultBackend));
```

Comment:
I think there might be an issue here - recapping my understanding of what this was doing before:

- if `dispatch_key` is an autograd key, `getBackendKeySetFromAutograd(dispatch_key)` will return the set of backend keys it maps to, otherwise the empty keyset.

- `hasKernelForDispatchKeySet(<keyset>)` returns true if we've registered a kernel for any of the keys in the set

...so the meaning of `is_autograd_key_with_backend_kernel` is, "`dispatch_key` is an autograd key and it maps to at least one backend key we have a kernel for".

So if that's accurate, then I think adding `DefaultBackend` like this changes the logic, since `is_autograd_key_with_backend_kernel` will now be true if we have a kernel registered for any backend key at all, regardless of whether `dispatch_key` is an autograd key or not, right?

Reply:
@bhosmer Note that we passed `DefaultBackend` to `hasKernelForAnyDispatchKey` instead of its Runtime keys. So `is_autograd_key_with_backend_kernel` will be true when there's a kernel registered to `DefaultBackend` or any backend in `autograd_other_backends` ;)

Reply:
Ah you're right, I was thinking about the alias set :P but still, say we have a kernel registered to `DefaultBackend` directly. Won't `is_autograd_key_with_backend_kernel` now be true for any value of `dispatch_key` at all? 

I'm not saying this necessarily breaks the logic where `is_autograd_key_with_backend_kernel` is used (I'm assuming it doesn't, since tests pass etc. :D) , but if my understanding is right, I think we should at least rename the variable (or even split it).

Reply:
> say we have a kernel registered to DefaultBackend directly. Won't is_autograd_key_with_backend_kernel now be true for any value of dispatch_key at all?
yep for any *Autograd* keys this will be true if there's a DefaultBackend. But I feel this kinda fits the current name `is_autograd_key_with_backend_kernel` where we just think of `DefaultBackend` as direct registration to all backends?

Reply:
sorry to be dense :) but what I'm not seeing is why this wouldn't be true for all dispatch keys, not just autograd keys. Like say dispatch_key == CPU. Then getBackendKeySetFromAutograd(dispatch_key).add(DispatchKey::DefaultBackend) would be {DispatchKey::DefaultBackend}, right?

The only other thing I can think of is that there are invariants that prevent non-autograd dispatch_key values from ever reaching this code, but I'm not sure I see how those would work... 

Anyway I'm sure I'm just missing something basic. You can set me straight tmr when we chat :)

Reply:
It's guarded downstream by `isIncludedInAlias(dispatch_key, DispatchKey::Math)` and `isIncludedInAlias(dispatch_key, DispatchKey::Autograd)`. Agree that the structure is confusing though.